### ISSUE TITLE
Baytrail: sound output is garbled with beating noise

### ISSUE THREAD

**Issue by plbossart** [thesofproject/linux#933]
I seriously question the results of the SOF CI and platform owners.

on Baytrail, a simple listening test on Asus T100 shows the audio is just corrupted, with beating noises and frequency shifts. This is just unacceptable, sorry, SOF is just complete garbage on all legacy platforms at this point. it's embarrassing to be that bad.

firmware 6eb79af18c7c8d6cefa4b8c80781fdeafd40adbb
kernel 35d7759

**Comment by fredoh9** [comment]
I have a Baytrail MinnowBoard(MB) i can take a look at.

**Comment by plbossart** [comment]
there's been some issues before like this with DMA pointers being in the weeds, can you try and grab the SSP signals with a LogicPro, then export the values in an .csv file. It's the best way to see if the samples play in sequence.

**Comment by fredoh9** [comment]
As i have only MB nocodec mode, i wanted to try to use loopback first but recording path has only silence. Not sure it was working before.
I need to check i can borrow LogicPro.

**Comment by juimonen** [comment]
@plbossart @fredoh9 is this still valid? I see other people and QA reporting that sound is ok, but there's timeout error with second playback... I also see this second playback issue in my logs, but the thing is that I have also crap sound, but now it seems that actually my hw is borked. I'm just confused about mixed reports here...

**Comment by wenqingfu** [comment]
@Jiangxinx reported that he cannot reproduce this issue on MinnowBoard+RT5651, see https://github.com/thesofproject/sof/pull/1451#issuecomment-494315890

**Comment by juimonen** [comment]
@plbossart I got the mb+rt5682 running. Well, sound is personally to me little different than hifiberry+up2, more high end emphasis at least. However I can't hear any obvious flaws, attached full scale sine sweep playback through mb+rt5682 and recording with my desktop.

**Comment by juimonen** [comment]
[sweep_mb_rt5682.zip](https://github.com/thesofproject/linux/files/3215540/sweep_mb_rt5682.zip)

**Comment by plbossart** [comment]
On 5/24/19 2:08 AM, juimonen wrote:
> @plbossart <https://github.com/plbossart> I got the mb+rt5682 running. 
> Well, sound is personally to me little different than hifiberry+up2, 
> more high end emphasis at least. However I can't hear any obvious flaws, 
> attached full scale sine sweep playback through mb+rt5682 and recording 
> with my desktop.

can you share kernel+firmware SHA1, I'll retry.

**Comment by juimonen** [comment]
@plbossart I think I was running it with several kernels like:
4d1fc20e3e01f7d91d9414f0265c7342b09931cf
c67168138451a041f1d07a4200890bb8cd43c4d2
0f3e63c8f1bffda8292fbcf3998c9a9bc66fa1a3
for me the top most was topic/sof-dev head (might have changed...)
I also used your latest broonie PR.

for fw I used like:
56b8776c2ace20cd0881c820ccecca810c80be12
which i think is master, but I also used some other 
versions including revert commit of dw-dma from Guennadi: 
(1da390743dade08da0478305f4bb5432971d443b)

So I would say just use latest/greatest from both. I can also 
test the versions you find problems with. Sorry I can't be more 
specific...

**Comment by plbossart** [comment]
No change on my side on ASUS T100TA (rt5640)
The sine wave looks really bad and my theory that the samples are played in the wrong order is likely still true.
![sine](https://user-images.githubusercontent.com/1149035/58358696-9e963c80-7e45-11e9-9252-73f70a93d39a.png)
[sine-asust100.wav.gz](https://github.com/thesofproject/linux/files/3218731/sine-asust100.wav.gz)

**Comment by plbossart** [comment]
binary used, @juimonen can you try to see if we might have a compilation issue?

[sof-byt.ri.gz](https://github.com/thesofproject/linux/files/3218741/sof-byt.ri.gz)

**Comment by plbossart** [comment]
Things work fine with the SST driver so it's not my device which is broken!

[sine-asust100-sst.wav.gz](https://github.com/thesofproject/linux/files/3218755/sine-asust100-sst.wav.gz)

**Comment by plbossart** [comment]
firmware a26c51a43e436a8da15fdad39d61e13bd4cdf12f
kernel 072cf06e35eb593a6694a30ac7aeb6b5c5d416a5

**Comment by juimonen** [comment]
@plbossart your binary works fine with my setup...

**Comment by juimonen** [comment]
I compiled this from scratch:
kernel: 924347de73e488fa21772af90566821694dcce55 (1 commit ahead of yours)
kernel config: config-daily-20190527.txt
fw: a26c51a43e436a8da15fdad39d61e13bd4cdf12f
tplg: sof-byt-rt5682.tplg

Again I can play fine in mb+rt5682... I'm puzzled what could be the issue with your t100

**Comment by juimonen** [comment]
@plbossart when playing in byt to hw0,1 (media pipeline with src) I get funny 8x speed effect. I can fix it with changing the pipe parameters to same as apl media pipe has (0,4 in apl/up2). However then I get buzzing effect with 44.1->48. 48->48 to 0,i in byt is clean. So I can just guess that there's some issue with dw-dma and 44.1 alternating frame sizes. Will try to debug it further. however as said before hw0,0 playback is clean for me in byt.

**Comment by plbossart** [comment]
@juimonen that helps progress even if we haven't root-caused the issue. 

it's not the topology file (same), not the firmware file (no difference), not the platform driver. So it's got to be codec driver or machine driver (I have the same issue with rt5645 on a CHT device). Can you try and see if there is any Asus T100 left in Espoo? Or are there any rt5651 devices left?

Also do you use the MCLK or not on the rt5682?

**Comment by singalsu** [comment]
On my Asus T100 the 16 bit playback is clean and OK. 24 bit and 32 bit content playback makes rattle (several times per second - I'll do a capture example and post the wavs and screenshots). 

This is with low latency pipeline. Media pipeline is complete garbage with >2x speedup in playback time with 16 bit content. With 24 and 32 bit content the media pipeline fails immediately without anything played.

**Comment by juimonen** [comment]
@plbossart what format are you using for playback?

So currently we have several issues found:
1) pcm hw:0,0 is good with 16 bit playback
2) pcm hw:0,0 is not good with s24/s32 playback (@singalsu reported above, I confirmed)
3) pcm hw:0,1 is not running at all or huge speed up audio
    3.1 there's fix to make src to work: https://github.com/thesofproject/sof/pull/1478
    3.2 to get rid of "speed metal" effect you need to copy apl media pipe params to byt (no PR yet)
4) After 3.1 & 3.2 hw:0,1 48khz is clean, but 44.1 conversion has ripple

So to sum up:
1) something strange in 0,0 with s24/s32, this is new thing, has to be debugged
2) 0,1 media pipe need several fixes, after fixes has still rattling with 44.1kHz

btw, me and @singalsu have quite uniform results in mb/rt5682 and t100.

**Comment by singalsu** [comment]
Here's my recordings from T100. 
[t100_wavs.zip](https://github.com/thesofproject/linux/files/3231970/t100_wavs.zip)

And plots of seen glitches (s16 clean, s24 and s32 format have similar issue)
![t100_s16](https://user-images.githubusercontent.com/28754793/58549880-2fd62c00-8215-11e9-9cca-34255e797db8.png)
![t100_s24](https://user-images.githubusercontent.com/28754793/58549894-382e6700-8215-11e9-87da-2c8ddb548255.png)
![t100_s32](https://user-images.githubusercontent.com/28754793/58549903-3d8bb180-8215-11e9-9d29-2ba3c165335c.png)

The glitches seem to be 1 ms long copies of previous audio, from about 20 ms ago. The glitches happen about every 32 ms.

**Comment by singalsu** [comment]
One more picture from a sound editor:

![Screenshot from 2019-05-29 13-32-57](https://user-images.githubusercontent.com/28754793/58550446-66f90d00-8216-11e9-903b-e2ee87227727.png)

**Comment by juimonen** [comment]
@plbossart I made 2 issues:
https://github.com/thesofproject/sof/issues/1482
https://github.com/thesofproject/sof/issues/1487

These seems to be quite the same with me and @singalsu in mb/rt5682 and t100.
So we haven't been yet able to produce similar waveform as you...?
These both seem to be "buffer wrap" issues, but yours looks more like harmonic distortion etc.

**Comment by singalsu** [comment]
@plbossart Could you try with this chirp. The corruptions are better visible in low frequency sine wave. Though your clip sounded like totally messed up and different to my T100 so not sure if this would reveal anything.

[chirp_20_20k_20s_log.wav.gz](https://github.com/thesofproject/linux/files/3232182/chirp_20_20k_20s_log.wav.gz)

**Comment by plbossart** [comment]
When I play a saw-tooth pattern on a MinnowBoard and grab the results with Logic Pro, I clearly see a discontinuity and as usual it's a step of 96 samples. Please redo the measurements by checking the signals on the connector first.
aplay -Dhw:0,0 -c2 -r48000 -f dat cnt_1024_left.pcm
[cnt_1024_left.pcm.gz](https://github.com/thesofproject/linux/files/3254087/cnt_1024_left.pcm.gz)



![ssp-gap](https://user-images.githubusercontent.com/1149035/58905898-6fe35600-86d0-11e9-86e4-0be0d2e0bbd1.png)
[ssp-gap.xlsx](https://github.com/thesofproject/linux/files/3254085/ssp-gap.xlsx)

**Comment by plbossart** [comment]
I have the issue issue on Asus 100 TA (rt5640), Asus T100HA (rt5645) and the Dialog DA7213 test board. All of them work fine with the closed-source firmware, none of them work with SOF.

**Comment by juimonen** [comment]
@plbossart I'll try again tomorrow in the office. btw did you have this https://github.com/thesofproject/sof/pull/1514 in your fw code? it fixed the s24/s32 playback for me...

**Comment by plbossart** [comment]
@juimonen yes this is with the latest firmware, but I play in S16_LE on hw:0,0 so your changes probably don't matter.

**Comment by singalsu** [comment]
Here's my analog recording from the Asus T100 (model T100TA-DK024H) with Pierre's file played with command:

`aplay -Dhw:0,0 -c2 -r48000 -f dat cnt_1024_left.pcm
`
The recording looks clean to me, no such repeating discontinuity as in above logic analyzer screenshot or in the spreadsheet data. I have fresh kernel, topology (sof-byt-rt5640.tplg), firmware fetched and built today.

[rec_cnt_1024_left_pcm.wav.gz](https://github.com/thesofproject/linux/files/3261692/rec_cnt_1024_left_pcm.wav.gz)

**Comment by singalsu** [comment]
Continued... if I had this corruption (plot made from Pierre's spreadsheet) it would show up well in the analog waveform too.

![ssp-gap](https://user-images.githubusercontent.com/28754793/59041470-5af9f480-8881-11e9-912e-2df0bc0f7982.png)

**Comment by juimonen** [comment]
@plbossart I'm also recording from analog output of codec and drawing from recorded wav-file. This kind of gaps should show up through the codec... I'm puzzled what you have there...

**Comment by plbossart** [comment]
I don't have an answer but can you please try and look at the connector level? You are using different codecs in different configurations so we don't have a common environment to test. The connector is the only common part and it should be the first test we do.

**Comment by juimonen** [comment]
@plbossart you are probably busy with other stuff, but is this still valid? If you have spare time to re-check with your setup... I think there has been some changes to byt ipc and other parts of the code.

**Comment by plbossart** [comment]
@juimonen ack, i'll retry tomorrow.

**Comment by plbossart** [comment]
not reproducible on ASUS T100 HAN (CHT), but PulseAudio speaker test generates audible underflows

**Comment by plbossart** [comment]
Not reproducible on Asus T100 TA (BYT) either. I wasn't able to check Pulseaudio, @juimonen can you give it a try with the UI speaker test?

**Comment by juimonen** [comment]
@plbossart in byt-mb aplay works fine but with pulseaudio I have similar issues than you, so quite nasty gaps in audio. Manually running pulseaudio doesn't spit out any xruns, also fw logger seems clean. 
With aplay I also get like small part of the audio repeated in the end when I cut the playback with ctrl-c... don't know if it is related, but seems this is not underrun, but some error in reporting the position or writing to audio buffer.

**Comment by juimonen** [comment]
btw, if I play with "paplay -vvvv" (verbose mode) the audio is clean (!?). And I see from dmesg position reporting that there's some difference. So with -vvv the buffer fill sizes are calculated somehow differently. Now the question is, why pulse "default" playback is not working, pulse or sof issue?

**Comment by plbossart** [comment]
it's definitively SOF, this doesn't happen with the legacy driver. I'll file another issue on this to avoid misunderstandings.

**Comment by plbossart** [comment]
follow-up on https://github.com/thesofproject/linux/issues/1285
